File: stanford-algo-course/go/week1/mergeSort.go

package main

import (
	"sort"	
)

const parallelismCutoff int = 30000

//Sort integer array in place. Function runs in n*log(n) time and n space.
func MergeSort(n []int) {
	if !sort.IntsAreSorted(n) {
		buffer := make([]int, len(n))
		copy(buffer, n)
		mergeSort(n, buffer)	
	}	
}

func mergeSort(toSort []int, buffer []int) {
	size := len(toSort)	
	if size < insertionSortCutoff {
		InsertionSort(toSort)
	} else {
		mid := size >> 1
		mergeSort(buffer[:mid], toSort[:mid])
		mergeSort(buffer[mid:], toSort[mid:])
		merge(buffer[:mid], buffer[mid:], toSort)
	}
}

func merge(left []int, right []int, buffer []int) {
	if left[len(left)-1] < right[0] { //optimized merge when alredy sorted
		copy(buffer, left)
		copy(buffer[len(left):], right)
		return
	}
	leftIdx, rightIdx, idx := 0, 0, 0
	for idx < len(buffer) {
		if left[leftIdx] < right[rightIdx] {
			buffer[idx] = left[leftIdx]
			leftIdx++
		} else {
			buffer[idx] = right[rightIdx]
			rightIdx++
		}
		idx++
		if leftIdx >= len(left) {
			copy(buffer[idx:], right[rightIdx:])
			return
		}
		if rightIdx >= len(right) {
			copy(buffer[idx:], left[leftIdx:])
			return
		}
	}
}

//Run insertion sort on given slice
func InsertionSort(toSort []int) {
	for i := 1; i < len(toSort); i++ {
		x := toSort[i]
		j := i
		for j > 0 && toSort[j-1] > x {
			toSort[j] = toSort[j-1]
			j--
		}
		toSort[j] = x
	}
}

func ParallelSort(n []int) {	
	buffer := make([]int, len(n))
	copy(buffer, n)
	parallelMergeSort(n, buffer, nil)
}

func parallelMergeSort(toSort []int, buffer []int, done chan bool) {	
	size := len(toSort)	
	if (size < parallelismCutoff) {
		MergeSort(toSort)
		if (done != nil) {
			done <- true
		}		
	} else {
		mid := size >> 1
		childDone := make(chan bool)
		go parallelMergeSort(buffer[:mid], toSort[:mid], childDone)
		parallelMergeSort(buffer[mid:], toSort[mid:], nil)
		<- childDone
		merge(buffer[:mid], buffer[mid:], toSort)	
		if (done != nil) {
			done <- true
		}
	}
}

File: stanford-algo-course/java/src/main/java/optimized/MergeSort.java
package optimized;

import java.util.Arrays;

class MergeSort {
	
	private static final int INSERTION_SORT_CUTOFF = 32;

	/** Sort integer array in place. Function runs in n*log(n) time and n space.*/
	public void sort(int[] input) {
		if (Sorts.isSorted(input)) return;
		int[] buffer = Arrays.copyOf(input, input.length);
		mergeSort(input, 0, input.length, buffer);
	}
	
	void mergeSort(int[] input, int start, int end, int[] buffer) {
		final int size = end - start;
		if (size < 2) return;
		if (size < INSERTION_SORT_CUTOFF) {
			InsertionSort.sort(input, start, end);
		} else {
			final int mid = (start + end) >>> 1;
			mergeSort(buffer, start, mid, input);
			mergeSort(buffer, mid, end, input);
			merge(buffer, start, mid, end, input);
		}
	}

	void merge(final int[] src, final int start, final int mid, final int end, final int[] dest) {
		if (src[mid-1] < src[mid]) {//optimized merge when already sorted
			System.arraycopy(src, start, dest, start, end-start);
			return;
		}
		int leftIdx = start;
		int rightIdx = mid;
		for (int idx = start; idx < end; idx++) {
			dest[idx] = src[leftIdx] < src[rightIdx] ? src[leftIdx++] : src[rightIdx++]; 
			if (leftIdx >= mid){
				System.arraycopy(src, rightIdx, dest, idx+1, end-rightIdx);
				return;
			}
			if (rightIdx >= end){
				System.arraycopy(src, leftIdx, dest, idx+1, mid-leftIdx);
				return;
			}
		}
	}
}